stdlib/strtoul.c

```diff
@@ -5,58 +5,99 @@
  *
  * strtoul, strtol
  *
- * Copyright 2017-2018 Phoenix Systems
- * Author: Jakub Sejdak, Pawel Pisarczyk, Michal Miroslaw
+ * Copyright 2017-2018, 2023 Phoenix Systems
+ * Author: Jakub Sejdak, Pawel Pisarczyk, Michal Miroslaw, Gerard Swiderski
  *
  * This file is part of Phoenix-RTOS.
  *
  * %LICENSE%
  */
 
+#include <errno.h>
 #include <stdlib.h>
 #include <ctype.h>
+#include <limits.h>
 
 
 unsigned long int strtoul(const char *nptr, char **endptr, int base)
 {
-	unsigned long int t, v = 0;
+	unsigned long int cutoff, t, result = 0;
+	int cutlim, width = 0, negative = 0;
+	const char *sptr = nptr;
 
-	if ((base == 16 || base == 0) && nptr[0] == '0' && (nptr[1] | 0x20) == 'x') {
+	while (isspace(*sptr) != 0) {
+		sptr++;
+	}
+
+	if (*sptr == '-') {
+		negative = 1;
+		sptr++;
+	}
+	else if (*sptr == '+') {
+		sptr++;
+	}
+
+	if ((base == 16 || base == 0) && sptr[0] == '0' && (sptr[1] | 0x20) == 'x') {
 		base = 16;
-		nptr += 2;
-	} else if (base == 0 && nptr[0] == '0')
+		sptr += 2;
+	}
+	else if (base == 0 && sptr[0] == '0') {
 		base = 8;
-	else if (base == 0)
+	}
+	else if (base == 0) {
 		base = 10;
+	}
+
+	if (base < 2 || base > (11 + 'z' - 'a')) {
+		errno = EINVAL;
+		return 0;
+	}
+
+	cutoff = (unsigned long)(ULONG_MAX) / (unsigned long)base;
+	cutlim = (unsigned long)(ULONG_MAX) % (unsigned long)base;
 
-	for (; isdigit(*nptr) || isalpha(*nptr); ++nptr) {
-		t = *nptr - '0';
-		if (t > 9)
-			t = (*nptr | 0x20) - 'a' + 10;
+	while (isdigit(*sptr) || isalpha(*sptr)) {
+		t = *sptr - '0';
+		if (t > 9) {
+			t = (*sptr | 0x20) - 'a' + 10;
+		}
 
-		if (t >= (unsigned long int)base)
+		if (t >= (unsigned long int)base) {
 			break;
+		}
 
-		v = (v * base) + t;
+		if (t > cutoff || (result == cutoff && t > cutlim)) {
```

Comment:
Hmm, I can't imagine a condition when t would be bigger than cutoff

Reply:
Good catch, should be `result` instead of `t`.